### ISSUE TITLE
EGVRecord testNum

### ISSUE THREAD

**Issue by jmimu** [DexcTrack/dexctrack#6]
Using G4, there is an error in **EGVRecord.testNum()**: there is no **self.data[5]**.

It may have consequences with **insert_cal_sql** and when testing 16777215 for calibration events when filling **calibList**.

**Comment by DexcTrack** [comment]
Good catch. I've merged in an update to the G4 version of testNum() in database_records.py. As you pointed out, we cannot use the testNum value to identify Calibration Events on a G4 system. I don't have access to a G4 device, so it will be difficult for me to find an alternative. In database_records.py (which I borrowed from the dexcom_reader project), there is a Calibration class, so maybe there are corresponding records on a G4 which could be accessed.

Could you run a test for me on your G4 device? Try running 'python readdata.py'. On my G6, it looks like this:

```
> python readdata.py
Firmware.ProductId = OrionReceiver
Found Dexcom Orion Receiver S/N: PL75130020
Transmitter paired: 8GBXBP
Battery Status: CHARGING (100%)
Record count:
- Meter records: 110
- CGM records: 8759
- CGM commitable records: 8759
- Event records: 622
- Insertion records: 23
- Calibration records: 0
- User Setting Records: 243
```

On my G5 which later became a G6, I have always found 0 Calibration records. I'd like to know if you see more on your G4. If you have such records, then we could try to figure out whether they could be used as an alternate means to identify Calibration Events.

**Comment by jmimu** [comment]
Thank you for the update.
I can't access the G4 right now but with the following code:
`curs.execute('CREATE TABLE IF NOT EXISTS Calib( sysSeconds INT PRIMARY KEY, dispSeconds INT, meterSeconds INT, type INT, glucose INT, testNum INT, xx INT);')
insert_cal_sql = '''INSERT OR IGNORE INTO Calib( sysSeconds, dispSeconds, meterSeconds, type, glucose, testNum, xx) VALUES (?, ?, ?, ?, ?, ?, ?);'''

respList = self.ReadRecords('METER_DATA')
for cal_rec in respList:
    #print ('raw_data =',' '.join(' %02x' % ord(c) for c in cal_rec.raw_data))
    #print ('Calib(', cal_rec.system_secs, ',', cal_rec.display_secs, ', ', cal_rec.meter_secs, ', ', cal_rec.record_type, ', ', cal_rec.calib_gluc, ',', cal_rec.testNum)
    if (self.rr_version == 'g5') or (self.rr_version == 'g6'):
        curs.execute(insert_cal_sql, (cal_rec.system_secs, cal_rec.display_secs, cal_rec.meter_secs, cal_rec.record_type, cal_rec.calib_gluc, cal_rec.testNum, cal_rec.xx))
    else:
        curs.execute(insert_cal_sql, (cal_rec.system_secs, cal_rec.display_secs, cal_rec.meter_secs, 0, cal_rec.calib_gluc, cal_rec.testNum, 0))
`

I got correct values in DB:

> "346223468"	"346284814"	"346223438"	"0"	"149"	"0"	"0"
"346278307"	"346339653"	"346278277"	"0"	"195"	"0"	"0"

Time and values are correct.

**Comment by DexcTrack** [comment]
I've implemented an alternative fix by adding methods xx_testNum(), xx(), and testNum() to the G4 version of MeterRecord() in database_records.py. Functionally, this will be equivalent to your suggested fix, but it eliminates the need to check self.rr_version in readReceiver.py. My change to database_records.py has been merged in.

**Comment by DexcTrack** [comment]
I've reworked the Calibration handling code in dexctrack.py to no longer have any dependency on **testNum**. This should make it work on G4 systems, which don't have real **testNum** fields in their **MeterRecord** instances, as well as G5 and G6 systems which do. In addition, swapping the order of the SELECT statements for **Calib** and **EgvRecord** provides a substantial speed increase, since there are far fewer **Calib** instances than **EgvRecord** instances.


I don't have access to a G4 receiver, so cannot verify these changes on a G4, but they are working fine on a G6.


If you have access to a G4, please pull in the latest versions of database_records.py and dexctrack.py and check whether User Calibrations are now properly graphed.

**Comment by jmimu** [comment]
Hi, it works well.
I sill have to add the fixes from issue #5 and pull request #4
